lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -2438,14 +2438,15 @@ uint64_t RefType::getMaxFieldID() const { return 0; }
 
 circt::hw::FieldIDTypeInterface
 RefType::getFinalTypeByFieldID(uint64_t fieldID) const {
-  assert(fieldID == 0);
-  return *this;
+  return getType().getFinalTypeByFieldID(fieldID - 1);
```

Comment:
How these work is unfortunately rather coupled to various methods/passes/abstractions in the code base, which is why RefType doesn't work this way already.

However since we do hoist them out immediately the resulting incorrect functionality of, e.g., `walkGroundTypes` (might have name wrong) is probably less of a concern than was imagined when thinking we needed these to work within larger aggregates throughout the pipeline.  See also notion of "ground" , and the dispatch methods in `FIRRTLTypes.cpp` such as `FIRRTLType::isGround`.

At the very least:

1. This should surely handle `fieldID == 0` still.
2. MaxFieldID for RefType's is zero (immediately above), non-zero fieldID's here are invalid unless you change that.
3. rootChildFieldID should match whatever is done here, after changing/updating this throughout the codebase (probably it's own PR?).

It's possible your use case would be better served by local notions of what your pass wants from a "FieldID" concept, something already done in a few places, until someone sits down and teases this tangle apart.

Reply:
I think ref types + fieldID is well defined, since it's just a depth-first walk of the type and all type-wrappers take 1 in that walk.  Essentially, from a fieldID perspective, `ref<Int>` == `bundle<a: Int>` in terms of the numbering of the int.

Reply:
Which is to say, fieldID is purely a type-structure concept, it has nothing to do with what the types mean.

Reply:
Okay, split this out and worked on updating passes/utilities, see #5795 .